### PR TITLE
add new case "uefi_boot_from_device"

### PR DESCRIPTION
**Pull request by XueqiangWei**
1. boot from ide-hd/ide-cd
2. boot from scsi-hd/scsi-cd
3. boot from virtio-blk
4. boot from usb-storage

ID: 1805245, 1805246, 1805270, 1805273, 1805282
Signed-off-by: Xueqiang Wei <xuwei@redhat.com>

### CODE REVIEW

**Comment by XueqiangWei** [comment]
(01/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.cdrom.ahci_cd.with_bootindex.q35: PASS (50.18 s)
 (02/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.cdrom.ahci_cd.with_specify_device.q35: PASS (50.68 s)
 (03/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.cdrom.scsi_cd.with_bootindex.q35: PASS (107.18 s)
 (04/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.cdrom.scsi_cd.with_specify_device.q35: PASS (50.31 s)
 (05/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.ahci_disk.with_bootindex.q35: PASS (49.88 s)
 (06/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.ahci_disk.with_specify_device.q35: PASS (49.91 s)
 (07/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.virtio_blk_disk.with_bootindex.q35: PASS (50.78 s)
 (08/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.virtio_blk_disk.with_specify_device.q35: PASS (49.56 s)
 (09/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.scsi_hd_disk.with_bootindex.q35: PASS (51.40 s)
 (10/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.scsi_hd_disk.with_specify_device.q35: PASS (49.85 s)
 (11/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.usb_storage_disk.usb_uhci.with_bootindex.q35: PASS (50.06 s)
 (12/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.usb_storage_disk.usb_uhci.with_specify_device.q35: PASS (49.54 s)
 (13/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.usb_storage_disk.usb_ehci.with_bootindex.q35: PASS (51.35 s)
 (14/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.usb_storage_disk.usb_ehci.with_specify_device.q35: PASS (49.86 s)
 (15/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.usb_storage_disk.usb_xhci.with_bootindex.q35: PASS (50.12 s)
 (16/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.usb_storage_disk.usb_xhci.with_specify_device.q35: PASS (49.86 s)

**Comment by XueqiangWei** [comment]
@zhencliu please help review it. Many thanks.

File: qemu/tests/cfg/uefi_boot_from_device.cfg [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    force_create_image_stg = yes
+    remove_image_stg = yes
+    variants:
+        - cdrom:
+            cdroms = "test"
+            cdrom_test = "/tmp/test.iso"
+            boot_entry_info = "Booting UEFI QEMU .*(QEMU CD-ROM|DVD-ROM QM)"
+            start_vm = no
+            dev_name = cdrom
+            variants:
```
Comment (zhencliu) [inline qemu/tests/cfg/uefi_boot_from_device.cfg]: what about move these variants out, like:
variants:
    - cdrom:
       ......
variants:
        - with_bootindex:
            bootindex_stg = 0
            cdrom:
                bootindex_stg = ''
                bootindex_test = 0
        - with_specify_device:
            cdrom:
                boot_dev = "UEFI QEMU (?:QEMU CD-ROM|DVD-ROM QM)"
         ......

File: qemu/tests/uefi_boot_from_device.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        """
+        return re.search(info, vm.logsessions['seabios'].get_output(), re.S)
+
+    def count_of_move_step(boot_dev):
+        """
+        get the number of steps to move
+        """
+        logs = vm.logsessions['seabios'].get_output()
+        for i in re.findall(boot_dev, logs, re.S):
+            return len(re.findall(r"Boot\d+:\sUEFI", logs.split(i)[0], re.S))
```
Comment (zhencliu) [inline qemu/tests/uefi_boot_from_device.py]: traverse a list,  return len(xxx) when handling the first item of the return of findall, so, why use 'for' loop here?

File: qemu/tests/uefi_boot_from_device.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    if dev_name == "cdrom":
+        cdrom_test = params["cdrom_test"]
+        create_cdroms(cdrom_test)
+        params["start_vm"] = "yes"
+        env_process.preprocess_vm(test, params, env, params.get("main_vm"))
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+
+    try:
```
Comment (zhencliu) [inline qemu/tests/uefi_boot_from_device.py]: 'try' to do something, should we have to handle exception if any occurs?

File: qemu/tests/cfg/uefi_boot_from_device.cfg (hunk 1 above)
Comment (XueqiangWei) [reply qemu/tests/cfg/uefi_boot_from_device.cfg]: I think no need to update. It's more clearer. thanks.

File: qemu/tests/uefi_boot_from_device.py (hunk 2 above)
Comment (XueqiangWei) [reply qemu/tests/uefi_boot_from_device.py]: you mean use re.findall(boot_dev, logs, re.S)[0]. The return value is a list, so I think they are all OK. I will update it.

File: qemu/tests/uefi_boot_from_device.py (hunk 3 above)
Comment (XueqiangWei) [reply qemu/tests/uefi_boot_from_device.py]: don't handle exception, just cleanup cdroms finally. thanks

**Comment by XueqiangWei** [comment]
(01/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.cdrom.ahci_cd.with_bootindex.q35: PASS (50.19 s)
 (02/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.cdrom.ahci_cd.with_specific_device.q35: PASS (51.30 s)
 (03/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.cdrom.scsi_cd.with_bootindex.q35: PASS (51.16 s)
 (04/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.cdrom.scsi_cd.with_specific_device.q35: PASS (50.90 s)
 (05/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.ahci_disk.with_bootindex.q35: PASS (50.29 s)
 (06/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.ahci_disk.with_specific_device.q35: PASS (49.81 s)
 (07/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.virtio_blk_disk.with_bootindex.q35: PASS (50.24 s)
 (08/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.virtio_blk_disk.with_specific_device.q35: PASS (50.80 s)
 (09/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.scsi_hd_disk.with_bootindex.q35: PASS (51.38 s)
 (10/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.scsi_hd_disk.with_specific_device.q35: PASS (50.12 s)
 (11/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.usb_storage_disk.usb_uhci.with_bootindex.q35: PASS (50.28 s)
 (12/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.usb_storage_disk.usb_uhci.with_specific_device.q35: PASS (49.65 s)
 (13/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.usb_storage_disk.usb_ehci.with_bootindex.q35: PASS (50.34 s)
 (14/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.usb_storage_disk.usb_ehci.with_specific_device.q35: PASS (49.68 s)
 (15/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.usb_storage_disk.usb_xhci.with_bootindex.q35: PASS (50.10 s)
 (16/16) Host_RHEL.m8.u2.product_av.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.uefi_boot_from_device.usb_storage_disk.usb_xhci.with_specific_device.q35: PASS (50.06 s)

**Comment by XueqiangWei** [comment]
@zhencliu please help review it again. Many thanks.

**Comment by zhencliu** [comment]
LGTM

**Comment by XueqiangWei** [comment]
@xutian please help review it.Many thanks.

File: qemu/tests/cfg/uefi_boot_from_device.cfg [hunk 4]
```diff
@@ -0,0 +1,82 @@
+- uefi_boot_from_device:
+    virt_test_type = qemu
+    type = uefi_boot_from_device
+    only q35
+    only ovmf
```
Comment (xutian) [inline qemu/tests/cfg/uefi_boot_from_device.cfg]: only x86, x86_64 ?

**Comment by xutian** [comment]
ACK

**Pull request merged**